File: data-structures/functional/adjacencyList.js
// An Immutable Adjacency List Representation of a Graph with Weighted Edges and
// Structural Sharing

class AL {
  constructor(vs = []) {
    this.vs = vs
  }
}

class Vertex {
  constructor(v, es = []) {
    this.v = v
    this.es = es
  }
}

class Edge {
  constructor(x, y, w) {
    this.f = x
    this.t = y
    this.w = w
  }
}

// adjacent :: AL -> Vertex -> Vertex -> Bool
function adjacent(G, x, y) {
  if (G.vs.some(v => v === x) && G.vs.some(v => v === y))
    return x.es.some(e => e.t === y)
}

// adjacentVals :: AL a => AL -> a -> a -> Bool
function adjacentVals(G, xv, yv) {
  const x = G.vs.filter(v => v.v === xv)[0]
  if (G.vs.some(v => v.v === xv) && G.vs.some(v => v.v === yv))
    return x.es.some(e => e.t.v === yv)
}

// neighbors :: AL -> Vertex -> [Vertex]
function neighbors(G, x) {
  if (G.vs.some(v => v === x))
    return x.es.map(e => e.t)
}

// neighborsByVal :: AL a => AL -> a -> [Vertex]
function neighborsByVal(G, xv) {
  const x = G.vs.filter(v => v.v === xv)[0]
  if (x) {
    return x.es.map(e => e.t)
  }
}

// addVertex :: AL -> Vertex -> AL
function addVertex(G, x) {
  return new AL(G.vs.concat(x))
}

// addE :: AL -> Vertex -> Vertex -> Int -> AL
function addE(G, x, y, w) {
  const nVs = G.vs.filter(v => v !== x && v !== y),
    nY = new Vertex(y.v, y.es),
    nX = new Vertex(x.v),
    nXes = x.es.filter(e => e.t !== y).concat(new Edge(nX, nY, w)),
    nYes = nY.es.map(e => {
      if (e.t === x) {
        return new Edge(nY, nX, e.w)
      }
      return e
    })
  nX.es = nXes
  nY.es = nYes
  return new AL(nVs.concat(nX, nY))
}

// addEByVals :: AL a => AL -> a -> a -> Int -> AL
function addEByVals(G, xv, yv, w) {
  const nVs = G.vs.filter(v => v.v !== xv && v.v !== yv),
    y = G.vs.filter(v => v.v === yv)[0],
    x = G.vs.filter(v => v.v === xv)[0],
    nY = new Vertex(yv, y.es),
    nX = new Vertex(xv),
    nXes = x.es.filter(e => e.t !== y).concat(new Edge(nX, nY, w)),
    nYes = nY.es.map(e => {
      if (e.t === x) {
        return new Edge(nY, nX, e.w)
      }
      return e
    })
  nX.es = nXes
  nY.es = nYes
  return new AL(nVs.concat(nX, nY))
}

// removeVertex :: AL -> Vertex -> AL
function removeVertex(G, x) {
  const nVs = G.vs.filter(v => v !== x)
    .reduce((nvs, v) => {
      if (v.es.some(e => e.t === x)) {
        const nEs = v.es.filter(e => e.t !== x)
        return nvs.concat(new Vertex(v.v, nEs))
      }
      return nvs.concat(v)
    }, [])
  return new AL(nVs)
}

// delE :: AL -> Vertex -> Vertex -> AL
function delE(G, x, y) {
  const nX = new Vertex(x.v, x.es.filter(e => e.t !== y))
  return addVertex(removeVertex(G, x), nX)
}

// getEVal :: AL a => AL -> a -> a -> Int
function getEVal(G, xv, yv) {
  const x = G.vs.filter(v => v.v === xv)[0],
    y = G.vs.filter(v => v.v === yv)[0]
  return x.es.filter(e => e.t === y)[0].w
}

const log = console.log,
  vs = [1, 2, 3, 4, 5, 6].map(v => new Vertex(v))
const G = vs.reduce(addVertex, new AL()),
  G1 = addE(addE(addE(G, vs[0], vs[1], 10), vs[2], vs[3], 20), vs[4], vs[5], 30),
  G2 = addEByVals(addEByVals(G1, 2, 1, 15), 2, 1, 20)
log(getEVal(G2, 1, 2)) // 10
log(getEVal(addEByVals(G2, 1, 2, 7), 1, 2)) // 7

File: data-structures/functional/adjacencyList.hs
-- Adjacency List with Structural Sharing and Weighted Edges

data AL a = AL { vs :: [Vertex a] }
  deriving (Show)

data Vertex a = V { value :: a
                  , es :: [Edge a]
                  } deriving (Show)

data Edge a = E { from :: Vertex a
                , to :: Vertex a
                , getW :: Int
                } deriving (Show)

printAL (AL vs) = [(value v, [(value $ to e, getW e) | e <- es v]) | v <- vs]

belongs (AL vs) a = any ((== a) . value) vs

adjacent g@(AL vs) a b
  | belongs g a = any ((== b) . value . to) (es x)
  | otherwise   = False
  where x = head [x | x <- vs, value x == a]

neighbors :: Eq a => AL a -> a -> Maybe [Vertex a]
neighbors (AL vs) a
  | not $ null x  = Just [to e | e <- es (head x)]
  | otherwise     = Nothing
  where x = [y | y <- vs, value y == a]

addVertex :: Eq a => AL a -> Vertex a -> AL a
addVertex (AL vs) x = AL (x:vs)

addEdge :: Eq a => AL a -> a -> a -> Int -> AL a
addEdge (AL vs) a b w = AL (nX:nY:nVs)
  where
    nX = V a (nXE:[e | e <- es x, value (to e) /= b])
    nY = V b $
       foldr (\ e yes -> if value (to e) == a then (E nY nX (getW e)):yes else e:yes) [] $ es y
    nXE = E nX nY w
    nVs = [v | v <- vs, value v /= a, value v /= b]
    x = head [x' | x' <- vs, value x' == a]
    y = head [y' | y' <- vs, value y' == b]

v1 = (V 1 [])
v2 = (V 2 [])
g0 = foldl addVertex (AL []) [V x [] | x <- [1..6]]
g1 = addEdge (addEdge g0 2 3 30) 2 5 20
